### PR TITLE
Ghost Pool Protection Admin Button

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

You can disable certain ways ghosts can get back into the round to keep a living ghost pool for events to pull from. There is no option to prevent in round revives as that's not what this is here to prevent,, this should only be for ghosts creating new bodies to leave the pool, thus putting themselves out of the event.

## I JUST DISABLED EVERYTHING! MUAHAHAHA!
image

## Why It's Good For The Game

Admins need ghosts sometimes for events, so being able to cut off certain ghost sinks will help us out. I fucking hate running out of ghosts.

## Changelog
:cl:
admin: New admin button: Ghost Pool Protection! It lets you enable and disable which sources ghosts can rejoin the round, minus their physical bodies getting revived.
/:cl:

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
Dig the icons! Always nice to see more TGUI admin tools.

**Comment by CRITAWAKETS** [comment]
Not gonna lie if you need to prevent ghosts from using spawners or mafia because they are running away from your event that is probably a sign that you need to do better events

**Comment by MrDoomBringer** [comment]
@CRITAWAKETS i think this might be more in the case of planning an event but not wanting ghosts to all join ghostroles while you're still setting it up

File: code/modules/mob/living/brain/posibrain.dm [hunk 1]
```diff
@@ -48,6 +48,8 @@ GLOBAL_VAR(posibrain_notify_cooldown)
 /obj/item/mmi/posibrain/attack_self(mob/user)
 	if(!brainmob)
 		brainmob = new(src)
+	if(!(GLOB.ghost_role_flags & GHOSTROLE_STATION_SENTIENCE))
+		to_chat(user, "<span class='warning'>Central Command has outlawed posibrain sentience in this sector...</span>")
```
Comment (ATH1909) [inline code/modules/mob/living/brain/posibrain.dm:52]:
```suggestion
		to_chat(user, "<span class='warning'>Central Command has outlawed posibrain sentience in this sector.</span>")
```

**Comment by ATH1909** [comment]
Are you sure that we even should HAVE messages for these? IMO, they should just silently pretend that no ghosts wanted to sign up to be that posibrain/pAI/whatever. Otherwise, someone is going to post a screenshot of one of these messages on the r/ss13 subreddit with a title like "OMG /TG/STATION DISABLED POSIBRAINS WHAT WILL ORANGES REMOVE NEXT".

Er, to clarify, these current messages make it sound like these roles are permanently disabled by a config option or something, not just temporarily disabled for an admin event.

**Comment by BeeSting12** [comment]
It's clear that admins temporarily disabled it through the message. If you want to make it further OOC then feel free.

File: code/modules/clothing/head/helmet.dm [hunk 2]
```diff
@@ -375,6 +375,8 @@
 		something.apply_damage(5,BRUTE,BODY_ZONE_HEAD,FALSE,FALSE,FALSE) //notably: no damage resist (it's in your helmet), no damage spread (it's in your helmet)
 		playsound(src, 'sound/machines/buzz-sigh.ogg', 30, TRUE)
 		return
+	if(!(GLOB.ghost_role_flags & GHOSTROLE_STATION_SENTIENCE))
+		say("ERROR: Central Command has outlawed monkey sentience helmets in this sector. NEAREST LAWFUL SECTOR: 2.537 million light years away.")
```
Comment (MrDoomBringer) [inline code/modules/clothing/head/helmet.dm:379]:
```suggestion
		say("ERROR: Central Command has temporarily outlawed monkey sentience helmets in this sector. NEAREST LAWFUL SECTOR: 2.537 million light years away.")
```

File: code/modules/awaymissions/corpse.dm [hunk 3]
```diff
@@ -37,6 +37,9 @@
 /obj/effect/mob_spawn/attack_ghost(mob/user)
 	if(!SSticker.HasRoundStarted() || !loc || !ghost_usable)
 		return
+	if(!(GLOB.ghost_role_flags & GHOSTROLE_SPAWNER) && !(flags_1 & ADMIN_SPAWNED_1))
+		to_chat(user, "<span class='warning'>An admin has disabled non-admin ghost roles!</span>")
```
Comment (MrDoomBringer) [inline code/modules/awaymissions/corpse.dm:41]:
```suggestion
		to_chat(user, "<span class='warning'>An admin has temporarily disabled non-admin ghost roles!</span>")
```

File: code/modules/awaymissions/capture_the_flag.dm [hunk 4]
```diff
@@ -216,6 +216,9 @@
 			return
 
 
+		if(!(GLOB.ghost_role_flags & GHOSTROLE_MINIGAME))
+			to_chat(user, "<span class='warning'>CTF has been disabled by admins.</span>")
```
Comment (MrDoomBringer) [inline code/modules/awaymissions/capture_the_flag.dm:220]:
```suggestion
			to_chat(user, "<span class='warning'>CTF has been temporarily disabled by admins.</span>")
```

File: code/controllers/subsystem/pai.dm [hunk 5]
```diff
@@ -139,6 +139,9 @@ SUBSYSTEM_DEF(pai)
 	return FALSE
 
 /datum/controller/subsystem/pai/proc/findPAI(obj/item/paicard/p, mob/user)
+	if(!(GLOB.ghost_role_flags & GHOSTROLE_SPAWNER))
+		to_chat(user, "<span class='warning'>Due to growing incidents of SELF corrupted independent artificial intelligences, freeform personality devices have been banned in this sector.</span>")
```
Comment (MrDoomBringer) [inline code/controllers/subsystem/pai.dm:143]:
```suggestion
		to_chat(user, "<span class='warning'>Due to growing incidents of SELF corrupted independent artificial intelligences, freeform personality devices have been temporarily banned in this sector.</span>")
```

File: code/modules/mob/living/brain/posibrain.dm (hunk 1 above)
Comment (MrDoomBringer) [inline code/modules/mob/living/brain/posibrain.dm:52]:
```suggestion
		to_chat(user, "<span class='warning'>Central Command has temporarily outlawed posibrain sentience in this sector...</span>")
```

**Review by MrDoomBringer — CHANGES_REQUESTED** [review]
I feel like just adding the 'temporarily' qualifier makes this seem much less like a config option and more like something that is on a more per-round basis

**Comment by tralezab** [comment]
Eh bikeshedding to argue about the small fluff messages i put on a few types. Others (ghost polling candidates) just return with no candidates always

### reminder: what i actually wanna hear from you athath is what roles are still available with all ghost role types disabled!

**Comment by Couls** [comment]
Don't like the idea of being forced out of ghost roles and the like, especially if I don't want to participate in an admin event or wish to kill time as something like a slaughter demon while its being set up

**Comment by tralezab** [comment]
Admins should be careful with this with that in mind, but it's necessary because you couldn't pull off events you could in say, 2015 because the amount of ways to leave ghost pool.

**Comment by RaveRadbury** [comment]
This is really great, any chance that we could break down the toggles further? Higher granularity.

**Comment by tralezab** [comment]
What were you thinking of, Rave?

**Comment by interestingusernam3** [comment]
>someone is going to post a screenshot of one of these messages on the r/ss13 subreddit with a title like "OMG /TG/STATION DISABLED POSIBRAINS WHAT WILL ORANGES REMOVE NEXT".

and

**Comment by RaveRadbury** [comment]
> What were you thinking of, Rave?

An example would be something like 
-Silicons
-- posi
-- pAI

-Lavaland
--Ashwalkers
--Golems

So you could disable all silicon ghost roles or just one or the other. This might be out of the scope or spirit of the PR, but the nuance of it might prove to be beneficial?

**Comment by tralezab** [comment]
I'll make a category for sillycones since that seems pretty useful, but the other one is kinda "eh" since it's essentially the majority of mob spawns

File: code/modules/admin/verbs/ghost_pool_protection.dm [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	*/
+
+/datum/ghost_pool_menu/New(H)//H can either be a client or a mob due to byondcode(tm)
+	if (istype(H,/client))
+		var/client/C = H
+		holder = C //if its a client, assign it to holder
+	else
+		var/mob/M = H
+		holder = M.client //if its a mob, assign the mob's client to holder
+	new_role_flags = GLOB.ghost_role_flags
```
Comment (Rohesie) [inline code/modules/admin/verbs/ghost_pool_protection.dm:33]:
```suggestion
/datum/ghost_pool_menu/New(user)//user can either be a client or a mob due to byondcode(tm)
	if (istype(user, /client))
		var/client/user_client = user
		holder = user_client //if its a client, assign it to holder
	else
		var/mob/user_mob = user
		holder = user_mob.client //if its a mob, assign the mob's client to holder
	new_role_flags = GLOB.ghost_role_flags
```
* Better var names.

File: code/modules/admin/verbs/ghost_pool_protection.dm [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		ui.open()
+
+/datum/ghost_pool_menu/ui_data(mob/user)
+	var/list/data = list()
+	data["events_or_midrounds"] = (new_role_flags & GHOSTROLE_MIDROUND_EVENT)
+	data["spawners"] = (new_role_flags & GHOSTROLE_SPAWNER)
+	data["station_sentience"] = (new_role_flags & GHOSTROLE_STATION_SENTIENCE)
+	data["silicons"] = (new_role_flags & GHOSTROLE_SILICONS)
+	data["minigames"] = (new_role_flags & GHOSTROLE_MINIGAME)
+	return data
```
Comment (Rohesie) [inline code/modules/admin/verbs/ghost_pool_protection.dm:54]:
```suggestion
	. = list()
	.["events_or_midrounds"] = (new_role_flags & GHOSTROLE_MIDROUND_EVENT)
	.["spawners"] = (new_role_flags & GHOSTROLE_SPAWNER)
	.["station_sentience"] = (new_role_flags & GHOSTROLE_STATION_SENTIENCE)
	.["silicons"] = (new_role_flags & GHOSTROLE_SILICONS)
	.["minigames"] = (new_role_flags & GHOSTROLE_MINIGAME)

```

File: code/modules/admin/verbs/ghost_pool_protection.dm [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+/datum/ghost_pool_menu/ui_act(action, params)
+	if(..())
+		return
```
Comment (Rohesie) [inline code/modules/admin/verbs/ghost_pool_protection.dm:58]:
```suggestion
/datum/ghost_pool_menu/ui_act(action, params)
	. = ..()
	if(.)
		return
```

File: code/modules/admin/verbs/ghost_pool_protection.dm [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+			new_role_flags = ALL
+		if("no_roles")
+			new_role_flags = NONE
+		if("apply_settings")
+			to_chat(usr, "Settings Applied!")
+			var/msg = "modified"
+			if(new_role_flags == ALL)
+				msg = "enabled all of"
+			if(new_role_flags == NONE)
+				msg = "disabled all of"
```
Comment (Rohesie) [inline code/modules/admin/verbs/ghost_pool_protection.dm:80]:
```suggestion
			var/msg
			switch(new_role_flags)
				if(ALL)
					msg = "enabled all of"
				if(NONE)
					msg = "disabled all of"
				else
					msg = "modified"
```

File: code/modules/events/_event.dm [hunk 10]
```diff
@@ -53,6 +53,8 @@
 		return FALSE
 	if(EMERGENCY_ESCAPED_OR_ENDGAMED)
 		return FALSE
+	if(istype(typepath, /datum/round_event/ghost_role) && GHOSTROLE_MIDROUND_EVENT)
```
Comment (Rohesie) [inline code/modules/events/_event.dm:56]:
```suggestion
	if(ispath(typepath, /datum/round_event/ghost_role) && GHOSTROLE_MIDROUND_EVENT)
```
`typath` is never instantiated, is it?

File: code/modules/admin/verbs/ghost_pool_protection.dm (hunk 7 above)
Comment (tralezab) [reply code/modules/admin/verbs/ghost_pool_protection.dm:54]: we ended up working it out on discord, data is much nicer

**Comment by tralezab** [comment]
@Rohesie okay ready for merge! I think style already told me his complaints (bolding everything) and i got rid of that

**Pull request merged**